pkg/replicatedapp/api.go

```diff
@@ -40,8 +44,37 @@ type LicenseData struct {
 	License      *kotsv1beta1.License
 }
 
+func GetPreferredChannelSlug() (string, error) {
+	clientset, err := k8sutil.GetClientset()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get clientset")
+	}
+
+	namespace := util.AppNamespace()
+
+	configMap, err := clientset.CoreV1().ConfigMaps(namespace).Get(context.TODO(), types.KotsadmConfigMap, metav1.GetOptions{})
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get existing kotsadm config map")
+	}
+
+	slug, ok := configMap.Data["preferred-channel-slug"]
+	if !ok {
+		return "", nil
+	}
+
+	return slug, nil
+}
+
 func GetLatestLicense(license *kotsv1beta1.License) (*LicenseData, error) {
+	channelSlug, err := GetPreferredChannelSlug()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get preferred channel slug")
+	}
+
 	url := fmt.Sprintf("%s/license/%s", license.Spec.Endpoint, license.Spec.AppSlug)
+	if channelSlug != "" {
+		url = fmt.Sprintf("%s/%s", url, channelSlug)
```

Comment:
just to confirm, is the idea that this decision will be recorded by replicated.app and this channel will be used automatically by replicated.app for subsequent requests?

Reply:
@sgalsaleh i think the answer is yes. At least for requests to this endpoint, I'm not sure how this plays out in other areas. But hoping @marccampbell or maybe @divolgin can confirm. 

Over on the pending replicated.app pr from Siva [he mentioned](https://github.com/replicatedhq/vandoor/pull/5813/files/f26385692619093bd8f051825242d42784963cf6#r1667019824):

> We now check for the existence of a row in the license_instance table. If it exists, we use the channelId from that table to generate the license in L245. This prevents the downstream from switching channels once initialized and recorded on the license_instance table.

And does look like it actually does that. If there was a licenseInstance entry it will use that channel id instead:

```javascript
license = await stores.licenseStore.getLicenseFromIdAndChannelSlugId(licenseInstance.licenseId, licenseInstance.downstreamChannelId);
```

Reply:
Hard to comment on code that has been removed 😄 :
- [Handler](https://github.com/replicatedhq/vandoor/pull/5813/files/f26385692619093bd8f051825242d42784963cf6#diff-2665a6a6cd084fcdf907fb35a2e73b113fd435d3fa642c20f7cefa3155b93df2R191) still has channel slug in it while client does not?: `@Get("/:appSlug/:channelSlug")`
- Instance data is recoded asynchronously and may be delayed.
- When channel on a license changes, instance can make any number of requests before reporting new channel
    - There is also a distinction between what channel license is assigned to and what channel is currently deployed
- This should be tested e2e with a channel other than "stable" with a brand new instance